keps/sig-apps/5440-mutable-job-pod-resource-updates/README.md

```diff
@@ -0,0 +1,583 @@
+  disrupt running workloads and complicate resource management.
+- Allow mutating resource specifications of pods directly.
+- Allow mutating other job specifications beyond container resource requirements.
+- Support in-place pod resource updates (this is covered by separate KEPs).
+- Allow mutating of Pod Resources.
+- Allow mutating of ResourceClaims.
+
+## Proposal
+
+The proposal is to relax update validation for container resource specifications
+(CPU, memory, GPU, and extended resource requests and limits) in the pod template of suspended jobs.
+
+This change has minimal impact on the job-controller, as the job controller will
+use the updated resource specifications when creating new pods for the job.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to build a controller that implements job queueing with dynamic resource optimization.
+Users create v1.Job objects, and to control when the job can run, I have a webhook that
+forces the jobs to be created in a suspended state. The controller analyzes current
+cluster capacity and adjusts job resource requirements to optimize cluster utilization
+before unsuspending them.
+
+At job creation time, users may specify conservative resource estimates or may not know
+the optimal resource allocation for current cluster conditions. The queue controller can
+analyze available capacity, other queued jobs, and cluster utilization patterns to
+determine optimal CPU, memory, and GPU allocations. For example, it might adjust the number of GPUs based on current
+availability. By updating the job's resource requirements before unsuspending it, the
+controller ensures efficient resource utilization and better cluster throughput.
+
+#### Story 2
+
+As an user, I want to be able to submit a job to a queueing solution. If my job cannot be admitted due to quota limitations, it should be possible to change my request so that the workload will be admitted.
+
+#### Story 3
+
+As a cluster administrator, I want to monitor existing workloads and see how many resources they are actually using. If a workload is oversubscribed and their actual utilization is lower, I want to suspend that workload and checkpoint the workload via external APIs. I would then lower the request requirements to match the actual utilization and resubmit my job.
+
+### Risks and Mitigations
+
+- New API calls from queue controllers to update resource specifications. The mitigation
+  is for such controllers to make a single API call for both updating resources and
+  unsuspending the job.
+  
+- Potential for resource specification changes to make a job unschedulable if the
+  updated requirements exceed available cluster capacity. Queue controllers should
+  validate resource availability before making changes.
+
+- A race condition could theoretically happen if a job is unsuspended and then quickly
+  suspended again before resource updates, though this is not a typical use case pattern.
```

Comment:
I'm having trouble understanding this race condition. Since resource updates are only allowed on suspended jobs, the sequence of "unsuspend... before resource updates" seems contradictory. Could you clarify the scenario you had in mind?

Reply:
I don't think this is really a concern. I'll remove.